aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3V4RestSignerClient.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    if (null != authResponse) {
+      session = newSession(authResponse, session);
+    } else if (null != token()) {
+      session = newSession(token(), session);
+    }
+
+    return session;
+  }
+
+  @Value.Check
+  protected void check() {
+    Preconditions.checkArgument(
+        properties().containsKey(S3_SIGNER_URI) || properties().containsKey(CatalogProperties.URI),
+        "S3 signer service URI is required");
+  }
+
+  private Long tokenExpiresInMs() {
+    return PropertyUtil.propertyAsLong(
+        properties(),
+        OAuth2Properties.TOKEN_EXPIRES_IN_MS,
+        OAuth2Properties.TOKEN_EXPIRES_IN_MS_DEFAULT);
+  }
+
+  @Override
+  protected void processRequestPayload(
+      SdkHttpFullRequest.Builder mutableRequest,
+      byte[] signature,
+      byte[] signingKey,
+      Aws4SignerRequestParams signerRequestParams,
+      AwsS3V4SignerParams signerParams) {
+    checkSignerParams(signerParams);
+  }
+
+  @Override
+  protected void processRequestPayload(
+      SdkHttpFullRequest.Builder mutableRequest,
+      byte[] signature,
+      byte[] signingKey,
+      Aws4SignerRequestParams signerRequestParams,
+      AwsS3V4SignerParams signerParams,
+      SdkChecksum sdkChecksum) {
+    checkSignerParams(signerParams);
+  }
+
+  @Override
+  protected String calculateContentHashPresign(
+      SdkHttpFullRequest.Builder mutableRequest, Aws4PresignerParams signerParams) {
+    return UNSIGNED_PAYLOAD;
+  }
+
+  @Override
+  public SdkHttpFullRequest presign(
+      SdkHttpFullRequest request, ExecutionAttributes executionAttributes) {
+    throw new UnsupportedOperationException("Pre-signing not allowed.");
+  }
+
+  @Override
+  public SdkHttpFullRequest sign(
+      SdkHttpFullRequest request, ExecutionAttributes executionAttributes) {
+    AwsS3V4SignerParams signerParams =
+        extractSignerParams(AwsS3V4SignerParams.builder(), executionAttributes).build();
+
+    S3SignRequest remoteSigningRequest =
+        ImmutableS3SignRequest.builder()
+            .method(request.method().name())
+            .region(signerParams.signingRegion().id())
+            .uri(request.getUri())
+            .headers(request.headers())
+            .properties(requestPropertiesSupplier().get())
+            .build();
+
+    Key cacheKey = Key.from(remoteSigningRequest);
+    SignedComponent cachedSignedComponent = SIGNED_COMPONENT_CACHE.getIfPresent(cacheKey);
+
+    if (cachedSignedComponent != null) {
+      // The SdkHttpFullRequest Builder appends the raw path from the input URI in .uri(),
+      // so we need to clear the current path from the request
+      SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
+      mutableRequest.encodedPath("");
+      mutableRequest.uri(cachedSignedComponent.signedURI());
+      reconstructHeaders(cachedSignedComponent.headers(), mutableRequest);
+
+      return mutableRequest.build();
```

Comment:
This duplicates the handling below. Can you refactor to avoid the duplication? Something like this:

```java
  SignedComponent signed;
  if (cachedSignedComponent != null) {
    signed = cachedSignedComponent;
  } else {
    S3SignResponse s3SignResponse = httpClient().post(...)
  }

  SdkHttpFullRequest.Builder request ...
```

Reply:
good point, I've updated this part to make it easier to read